### PR TITLE
VIAC French import for PDF Extractor, basic tests

### PR DESCRIPTION
**Pull request by denispol**
Adding PDF Importer for WirBank documents (VIAC) in French (only English and German are supported in current release)

### CODE REVIEW

File: name.abuchen.portfolio.tests/src/name/abuchen/portfolio/datatransfer/pdf/wirbank/WirBankPDFExtractorTest.java [hunk 1]
```diff
@@ -807,6 +807,56 @@ public void testWertpapierKauf10WithSecurityInCHF()
...
+        new AssertImportActions().check(results, "CHF");
+
+        // check security
+        Security security = results.stream().filter(SecurityItem.class::isInstance).findFirst()
+                        .orElseThrow(IllegalArgumentException::new).getSecurity();
+        assertThat(security.getIsin(), is("CH0030849613"));
+        assertNull(security.getWkn());
+        assertNull(security.getTickerSymbol());
+        assertThat(security.getName(), is("CSIF Canada"));
+        assertThat(security.getCurrencyCode(), is(CurrencyUnit.CAD));
```
Comment (Nirus2000) [inline name.abuchen.portfolio.tests/src/name/abuchen/portfolio/datatransfer/pdf/wirbank/WirBankPDFExtractorTest.java:832]:
```suggestion
        assertThat(security.getCurrencyCode(), is("CAD"));
```

We have only defined two CurrencyUnit as fixed constants. Once EUR and once USD. All other currencies are to be set as strings.

File: name.abuchen.portfolio.tests/src/name/abuchen/portfolio/datatransfer/pdf/wirbank/WirBankPDFExtractorTest.java [hunk 2]
```diff
@@ -807,6 +807,56 @@ public void testWertpapierKauf10WithSecurityInCHF()
...
+        assertThat(entry.getPortfolioTransaction().getGrossValue(),
+                        is(Money.of("CHF", Values.Amount.factorize(32.84))));
+        assertThat(entry.getPortfolioTransaction().getUnitSum(Unit.Type.TAX),
+                        is(Money.of("CHF", Values.Amount.factorize(0.00))));
+        assertThat(entry.getPortfolioTransaction().getUnitSum(Unit.Type.FEE),
+                        is(Money.of("CHF", Values.Amount.factorize(0.00))));
+
+        Unit grossValueUnit = entry.getPortfolioTransaction().getUnit(Unit.Type.GROSS_VALUE)
+                        .orElseThrow(IllegalArgumentException::new);
+        assertThat(grossValueUnit.getForex(), is(Money.of(CurrencyUnit.USD, Values.Amount.factorize(42.12))));
```
Comment (Nirus2000) [inline name.abuchen.portfolio.tests/src/name/abuchen/portfolio/datatransfer/pdf/wirbank/WirBankPDFExtractorTest.java:857]:
```suggestion
        assertThat(grossValueUnit.getForex(), is(Money.of("CAD", Values.Amount.factorize(42.12))));
```

File: name.abuchen.portfolio.tests/src/name/abuchen/portfolio/datatransfer/pdf/wirbank/WirBankPDFExtractorTest.java [hunk 3]
```diff
@@ -912,6 +962,41 @@ public void testInterest03()
...
+
+        // check interest transaction
+        AccountTransaction transaction = (AccountTransaction) results.stream().filter(TransactionItem.class::isInstance)
+                        .findFirst().orElseThrow(IllegalArgumentException::new).getSubject();
+
+        assertThat(transaction.getType(), is(AccountTransaction.Type.INTEREST));
+
+        assertThat(transaction.getDateTime(), is(LocalDateTime.parse("2022-06-01T00:00")));
+        assertThat(transaction.getSource(), is("Zins06_Francais.txt"));
+        assertThat(transaction.getNote(), is("Taux d\\'intérêt: 0.10% | Période d\\'intérêt: mai"));
```
Comment (Nirus2000) [inline name.abuchen.portfolio.tests/src/name/abuchen/portfolio/datatransfer/pdf/wirbank/WirBankPDFExtractorTest.java:988]:
```suggestion
        assertThat(transaction.getNote(), is("Taux d'intérêt: 0.10% | Période d'intérêt: mai"));
```
We do not need to invalidate any strings or characters here.

**Comment by denispol** [comment]
Thank you for reviewing, I am not really knowledgeable in Java, made this PR to get VIAC imports working in French

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java [hunk 4]
```diff
@@ -64,8 +64,8 @@ private void addDepositTransaction()
                         // Gutschrift: Valuta 15.01.2019 CHF 2'150.00
                         // @formatter:on
                         .section("date", "amount", "currency") //
-                        .find("(Einzahlung|Deposit) 3a") //
-                        .match("^(Gutschrift: Valuta|Credit: Value date) (?<date>[\\d]{2}\\.[\\d]{2}\\.[\\d]{4}) (?<currency>[\\w]{3}) (?<amount>[\\.,'\\d]+)$") //
+                        .find("(Einzahlung|Deposit|Versement) 3a") //
+                        .match("^(Gutschrift: Valuta|Credit: Value date|Montant crédité: Valeur) (?<date>[\\d]{2}\\.[\\d]{2}\\.[\\d]{4}) (?<currency>[\\w]{3}) (?<amount>[\\.,'\\d]+)$") //
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java:68]:
```suggestion
                        .match("^(Gutschrift: Valuta|Credit: Value date|Montant cr.dit.: Valeur) (?<date>[\\d]{2}\\.[\\d]{2}\\.[\\d]{4}) (?<currency>[\\w]{3}) (?<amount>[\\.,'\\d]+)$") //
```
Special characters are to be devalued with a dot as a regular expression.

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java [hunk 5]
```diff
@@ -84,12 +84,12 @@ private void addDepositTransaction()
 
     private void addBuySellTransaction()
     {
-        DocumentType type = new DocumentType("(B.rsenabrechnung|Exchange Settlement) \\- (Kauf|Verkauf|Buy|Sell)");
+        DocumentType type = new DocumentType("(B.rsenabrechnung|Exchange Settlement|Opération de bourse) \\- (Kauf|Verkauf|Buy|Sell|Achat|Vente)");
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java:87]:
```suggestion
        DocumentType type = new DocumentType("(B.rsenabrechnung|Exchange Settlement|Op.ration de bourse) \\- (Kauf|Verkauf|Buy|Sell|Achat|Vente)");
```

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java [hunk 6]
```diff
@@ -84,12 +84,12 @@ private void addDepositTransaction()
...
         this.addDocumentTyp(type);
 
         Transaction<BuySellEntry> pdfTransaction = new Transaction<>();
 
-        Block firstRelevantLine = new Block("^(B.rsenabrechnung|Exchange Settlement) \\- (Kauf|Verkauf|Buy|Sell).*$");
+        Block firstRelevantLine = new Block("^(B.rsenabrechnung|Exchange Settlement|Opération de bourse) \\- (Kauf|Verkauf|Buy|Sell|Achat|Vente).*$");
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java:92]:
```suggestion
        Block firstRelevantLine = new Block("^(B.rsenabrechnung|Exchange Settlement|Op.ration de bourse) \\- (Kauf|Verkauf|Buy|Sell|Achat|Vente).*$");
```

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java [hunk 7]
```diff
@@ -103,9 +103,9 @@ private void addBuySellTransaction()
 
                         // Is type --> "Verkauf" change from BUY to SELL
                         .section("type").optional() //
-                        .match("^(B.rsenabrechnung|Exchange Settlement) \\- (?<type>(Kauf|Verkauf|Buy|Sell)).*$") //
+                        .match("^(B.rsenabrechnung|Exchange Settlement|Opération de bourse) \\- (?<type>(Kauf|Verkauf|Buy|Sell|Achat|Vente)).*$") //
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java:106]:
```suggestion
                        .match("^(B.rsenabrechnung|Exchange Settlement|Op.ration de bourse) \\- (?<type>(Kauf|Verkauf|Buy|Sell|Achat|Vente)).*$") //
```

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java [hunk 8]
```diff
@@ -116,33 +116,33 @@ private void addBuySellTransaction()
                         // Kurs: USD 262.51
                         // @formatter:on
                         .section("isin", "name", "currency") //
-                        .find("Order: (Kauf|Verkauf|Buy|Sell)") //
-                        .match("^[\\.,\\d]+ (Ant|Qty|Anteile|units) (?<name>.*)$") //
+                        .find("(Order|Ordre): (Kauf|Verkauf|Buy|Sell|Achat|Vente)") //
+                        .match("^[\\.,\\d]+ (Ant|Qty|Anteile|units|Qté|Quantité|parts|actions) (?<name>.*)$") //
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java:120]:
```suggestion
                        .match("^[\\.,\\d]+ (Ant|Qty|Anteile|units|Qt.|Quantit.|parts|actions) (?<name>.*)$") //
```

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java [hunk 9]
```diff
@@ -116,33 +116,33 @@ private void addBuySellTransaction()
...
+                        .find("(Order|Ordre): (Kauf|Verkauf|Buy|Sell|Achat|Vente)") //
+                        .match("^(?<shares>[\\.,\\d]+) (Ant|Qty|Qté|Anteile|units|parts) (?<name>.*)$") //
                         .assign((t, v) -> t.setShares(asShares(v.get("shares"))))
 
                         // @formatter:off
                         // Verrechneter Betrag: Valuta 05.07.2018 CHF 360.43
                         // @formatter:on
                         .section("date") //
-                        .match("^(Verrechneter Betrag: Valuta|Charged amount: Value date) (?<date>[\\d]{2}\\.[\\d]{2}\\.[\\d]{4}) .*$") //
+                        .match("^(Verrechneter Betrag: Valuta|Charged amount: Value date|Montant comptabilisé: Valeur) (?<date>[\\d]{2}\\.[\\d]{2}\\.[\\d]{4}) .*$") //
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java:138]:
```suggestion
                        .match("^(Verrechneter Betrag: Valuta|Charged amount: Value date|Montant comptabilis.: Valeur) (?<date>[\\d]{2}\\.[\\d]{2}\\.[\\d]{4}) .*$") //
```

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java [hunk 10]
```diff
@@ -116,33 +116,33 @@ private void addBuySellTransaction()
...
                         .assign((t, v) -> t.setDate(asDate(v.get("date"))))
 
                         // @formatter:off
                         // Verrechneter Betrag: Valuta 05.07.2018 CHF 360.43
                         // @formatter:on
                         .section("amount", "currency") //
-                        .match("^(Verrechneter Betrag: Valuta|Charged amount: Value date) .* (?<currency>[\\w]{3}) (?<amount>[\\.,'\\d]+)$") //
+                        .match("^(Verrechneter Betrag: Valuta|Charged amount: Value date|Montant comptabilisé: Valeur) .* (?<currency>[\\w]{3}) (?<amount>[\\.,'\\d]+)$") //
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java:145]:
```suggestion
                        .match("^(Verrechneter Betrag: Valuta|Charged amount: Value date|Montant comptabilis.: Valeur) .* (?<currency>[\\w]{3}) (?<amount>[\\.,'\\d]+)$") //
```

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java [hunk 11]
```diff
@@ -174,12 +174,12 @@ private void addBuySellTransaction()
 
     private void addInterestTransaction()
     {
-        DocumentType type = new DocumentType("(Zins|Interest)");
+        DocumentType type = new DocumentType("(Zins|Interest|Intérêts)");
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java:177]:
```suggestion
        DocumentType type = new DocumentType("(Zins|Interest|Int.r.ts)");
```

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java [hunk 12]
```diff
@@ -174,12 +174,12 @@ private void addBuySellTransaction()
...
         this.addDocumentTyp(type);
 
         Transaction<AccountTransaction> pdfTransaction = new Transaction<>();
 
-        Block firstRelevantLine = new Block("^(Zins|Interest)$");
+        Block firstRelevantLine = new Block("^(Zins|Interest|Intérêts)$");
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java:182]:
```suggestion
        Block firstRelevantLine = new Block("^(Zins|Interest|Int.r.ts)$");
```

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java [hunk 13]
```diff
@@ -197,9 +197,9 @@ private void addInterestTransaction()
                         // Zinsgutschrift: CHF 0.04
                         // @formatter:on
                         .section("date", "amount", "currency") //
-                        .find("(Zins|Interest)") //
-                        .match("^(Am|On) (?<date>[\\d]{2}\\.[\\d]{2}\\.[\\d]{4}) (haben wir (Ihrem Konto|Ihnen) gutgeschrieben|we have credited your account):$") //
-                        .match("^(Zinsgutschrift|Interest credit): (?<currency>[\\w]{3}) (?<amount>[\\.,'\\d]+)$") //
+                        .find("(Zins|Interest|Intérêts)") //
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java:200]:
```suggestion
                        .find("(Zins|Interest|Int.r.ts)") //
```

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java [hunk 14]
```diff
@@ -197,9 +197,9 @@ private void addInterestTransaction()
...
+                        .match("^(Am|On|Nous avons crédité le) (?<date>[\\d]{2}\\.[\\d]{2}\\.[\\d]{4}) (haben wir (Ihrem Konto|Ihnen) gutgeschrieben|we have credited your account|les intérêts suivants):$") //
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java:201]:
```suggestion
                        .match("^(Am|On|Nous avons cr.dit. le) (?<date>[\\d]{2}\\.[\\d]{2}\\.[\\d]{4}) (haben wir (Ihrem Konto|Ihnen) gutgeschrieben|we have credited your account|les int.r.ts suivants):$") //
```

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java [hunk 15]
```diff
@@ -197,9 +197,9 @@ private void addInterestTransaction()
...
+                        .match("^(Zinsgutschrift|Interest credit|Intérêts créditeurs): (?<currency>[\\w]{3}) (?<amount>[\\.,'\\d]+)$") //
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java:202]:
```suggestion
                        .match("^(Zinsgutschrift|Interest credit|Int.r.ts créditeurs): (?<currency>[\\w]{3}) (?<amount>[\\.,'\\d]+)$") //
```

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java [hunk 16]
```diff
@@ -211,8 +211,8 @@ private void addInterestTransaction()
                         // Zinsperiode: März
                         // @formatter:on
                         .section("note1", "note2").optional() //
-                        .match("^(?<note1>(Zinssatz|Interest rate): [\\.,\\d]+%)$") //
-                        .match("^(?<note2>(Zinsperiode|Interest period): .*)$") //
+                        .match("^(?<note1>(Zinssatz|Interest rate|Taux d'intérêt): [\\.,\\d]+%)$") //
+                        .match("^(?<note2>(Zinsperiode|Interest period|Période d'intérêt): .*)$") //
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java:215]:
```suggestion
                        .match("^(?<note1>(Zinssatz|Interest rate|Taux d.int.r.t): [\\.,\\d]+%)$") //
                        .match("^(?<note2>(Zinsperiode|Interest period|P.riode d.int.r.t): .*)$") //
```

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java [hunk 17]
```diff
@@ -243,7 +243,7 @@ private void addFeeTransaction()
                         // @formatter:on
                         .section("date", "amount", "currency") //
                         .find("(Belastung|Commission)") //
-                        .match("^(Verrechneter Betrag: Valuta|Charged amount: Value date) (?<date>[\\d]{2}\\.[\\d]{2}\\.[\\d]{4}) (?<currency>[\\w]{3}) (\\-)?(?<amount>[\\.,'\\d]+)$")
+                        .match("^(Verrechneter Betrag: Valuta|Charged amount: Value date|Montant débité: Valuta) (?<date>[\\d]{2}\\.[\\d]{2}\\.[\\d]{4}) (?<currency>[\\w]{3}) (\\-)?(?<amount>[\\.,'\\d]+)$")
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java:246]:
```suggestion
                        .match("^(Verrechneter Betrag: Valuta|Charged amount: Value date|Montant d.bit.: Valuta) (?<date>[\\d]{2}\\.[\\d]{2}\\.[\\d]{4}) (?<currency>[\\w]{3}) (\\-)?(?<amount>[\\.,'\\d]+)$")
```

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java [hunk 18]
```diff
@@ -254,7 +254,7 @@ private void addFeeTransaction()
                         // Effektive VIAC Verwaltungsgebühr: 0.123% p.a. CHF -1.11
                         // @formatter:on
                         .section("note").optional() //
-                        .match("^(Effektive|Effective) (?<note>(VIAC Verwaltungsgeb.hr|VIAC administration fee): [\\.,\\d]+%) .*$") //
+                        .match("^(Effektive|Effective|Commission de gestion effective) (?<note>(VIAC Verwaltungsgeb.hr|VIAC administration fee|VIAC p.a.): [\\.,\\d]+%) .*$") //
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java:257]:
```suggestion
                        .match("^(Effektive|Effective|Commission de gestion effective) (?<note>(VIAC Verwaltungsgeb.hr|VIAC administration fee|VIAC p\\.a\\.): [\\.,\\d]+%) .*$") //
```

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java [hunk 19]
```diff
@@ -271,17 +271,21 @@ private void addDividendTransaction()
     {
         DocumentType type = new DocumentType("(Dividendenaussch.ttung" //
                         + "|Dividend Payment" //
+                        + "|Avis de dividende"
                         + "|R.ckerstattung Quellensteuer" //
-                        + "|Refund withholding tax)");
+                        + "|Refund withholding tax" //
+                        + "|Remboursement de l'impôt à la source");
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java:277]:
```suggestion
                        + "|Remboursement de l.imp.t . la source");
```

File: name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java [hunk 20]
```diff
@@ -271,17 +271,21 @@ private void addDividendTransaction()
...
         Transaction<AccountTransaction> pdfTransaction = new Transaction<>();
 
         Block firstRelevantLine = new Block("^(Dividendenaussch.ttung" //
                         + "|Dividend Payment" //
                         + "|Cancelation Dividend Payment" //
+                        + "|Avis de dividende"
                         + "|R.ckerstattung Quellensteuer" //
-                        + "|Refund withholding tax)$");
+                        + "|Refund withholding tax" //
+                        + "|Remboursement de l'impôt à la source)$");
```
Comment (Nirus2000) [inline name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/WirBankPDFExtractor.java:288]:
```suggestion
                        + "|Remboursement de l.imp.t . la source)$");
```

**Comment by Nirus2000** [comment]
Hello @denispol 
Your pull request looks very good... there are a few small things that need to be adjusted.
But TOP !!! 👍🏻 👍🏻 👍🏻 

All special characters are to be replaced with a dot as a regular expression. So the { [ ( | ) ] } * + ? \ . ' ^ ... in the PDF document are to be escaped. (Escaping = \\.)

If you have time... check the rest of your source...

And that's all... VERY VERY GOOD... !! 

Regards Alex

**Comment by buchen** [comment]
Thanks @denispol for the contribution
and thank you very much to @Nirus2000 for shepherding the pull request into the source (#3633)

It is now merged.

**Pull request closed without merging**